// stmp/lib/messages.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-11-29 16:24:51
 *
 * FINE
 */
Object.defineProperty(exports, "__esModule", { value: true });
var MessageKind;
(function (MessageKind) {
    MessageKind["Ping"] = "Ping";
    MessageKind["Request"] = "Request";
    MessageKind["Push"] = "Push";
    MessageKind["Response"] = "Response";
    MessageKind["Follow"] = "Follow";
    MessageKind["Close"] = "Close";
})(MessageKind = exports.MessageKind || (exports.MessageKind = {}));
function Headers() {
    return Object.create(null);
}
exports.Headers = Headers;
function isPingMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Ping;
}
exports.isPingMessage = isPingMessage;
function isPayload(data, allowEmpty = true) {
    if (allowEmpty) {
        return (data === void 0 || typeof data === 'string' || data instanceof Uint8Array);
    }
    return typeof data === 'string' || data instanceof Uint8Array;
}
exports.isPayload = isPayload;
function isIncomeMessage(message) {
    return isRequestMessage(message) || isPushMessage(message);
}
exports.isIncomeMessage = isIncomeMessage;
function isRequestMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Request;
}
exports.isRequestMessage = isRequestMessage;
function isPushMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Push;
}
exports.isPushMessage = isPushMessage;
function isResponseMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Response;
}
exports.isResponseMessage = isResponseMessage;
function isCloseMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Close;
}
exports.isCloseMessage = isCloseMessage;
function isFollowMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Follow;
}
exports.isFollowMessage = isFollowMessage;
function isExchangeMessage(message) {
    return isIncomeMessage(message) || isResponseMessage(message);
}
exports.isExchangeMessage = isExchangeMessage;
exports.STATUS_UNKNOWN = 0;
exports.HTTP_OK = 200;
exports.HTTP_BAD_REQUEST = 400;
exports.HTTP_UNAUTHORIZED = 401;
exports.HTTP_NOT_FOUND = 404;
exports.HTTP_REQUEST_TIMEOUT = 408;
exports.HTTP_INTERNAL_SERVER_ERROR = 500;
exports.WS_NORMAL_CLOSURE = 1000;
exports.WS_PROTOCOL_ERROR = 1002;
exports.WS_ABNORMAL_CLOSURE = 1006;
exports.WS_INTERNAL_ERROR = 1011;
exports.WS_USER_MIN = 4000;
function createPingMessage(messageId) {
    return { kind: MessageKind.Ping, messageId };
}
exports.createPingMessage = createPingMessage;
function createRequestMessage(messageId, action, payload, headers = Headers(), fin = true) {
    return {
        kind: MessageKind.Request,
        payload,
        messageId,
        action,
        headers,
        fin,
    };
}
exports.createRequestMessage = createRequestMessage;
function createPushMessage(messageId, action, payload, headers = Headers(), fin = true) {
    return {
        kind: MessageKind.Push,
        payload,
        messageId,
        action,
        headers,
        fin,
    };
}
exports.createPushMessage = createPushMessage;
function createResponseMessage(messageId, status, payload, headers = Headers(), fin = true) {
    return {
        kind: MessageKind.Response,
        payload,
        status,
        messageId,
        headers,
        fin,
    };
}
exports.createResponseMessage = createResponseMessage;
function createCloseMessage(status, message) {
    return { kind: MessageKind.Close, message, status };
}
exports.createCloseMessage = createCloseMessage;
function createFollowMessage(messageId, payload, fin = true) {
    return { kind: MessageKind.Follow, messageId, payload, fin };
}
exports.createFollowMessage = createFollowMessage;
