// stmp/lib/messages.mjs
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-11-29 16:24:51
 *
 * FINE
 */
export var MessageKind;
(function (MessageKind) {
    MessageKind["Ping"] = "Ping";
    MessageKind["Request"] = "Request";
    MessageKind["Push"] = "Push";
    MessageKind["Response"] = "Response";
    MessageKind["Follow"] = "Follow";
    MessageKind["Close"] = "Close";
})(MessageKind || (MessageKind = {}));
export function Headers() {
    return Object.create(null);
}
export function isPingMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Ping;
}
export function isPayload(data, allowEmpty) {
    if (allowEmpty === void 0) { allowEmpty = true; }
    if (allowEmpty) {
        return (data === void 0 || typeof data === 'string' || data instanceof Uint8Array);
    }
    return typeof data === 'string' || data instanceof Uint8Array;
}
export function isIncomeMessage(message) {
    return isRequestMessage(message) || isPushMessage(message);
}
export function isRequestMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Request;
}
export function isPushMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Push;
}
export function isResponseMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Response;
}
export function isCloseMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Close;
}
export function isFollowMessage(message) {
    var _a;
    return ((_a = message) === null || _a === void 0 ? void 0 : _a.kind) === MessageKind.Follow;
}
export function isExchangeMessage(message) {
    return isIncomeMessage(message) || isResponseMessage(message);
}
export var STATUS_UNKNOWN = 0;
export var HTTP_OK = 200;
export var HTTP_BAD_REQUEST = 400;
export var HTTP_UNAUTHORIZED = 401;
export var HTTP_NOT_FOUND = 404;
export var HTTP_REQUEST_TIMEOUT = 408;
export var HTTP_INTERNAL_SERVER_ERROR = 500;
export var WS_NORMAL_CLOSURE = 1000;
export var WS_PROTOCOL_ERROR = 1002;
export var WS_ABNORMAL_CLOSURE = 1006;
export var WS_INTERNAL_ERROR = 1011;
export var WS_USER_MIN = 4000;
export function createPingMessage(messageId) {
    return { kind: MessageKind.Ping, messageId: messageId };
}
export function createRequestMessage(messageId, action, payload, headers, fin) {
    if (headers === void 0) { headers = Headers(); }
    if (fin === void 0) { fin = true; }
    return {
        kind: MessageKind.Request,
        payload: payload,
        messageId: messageId,
        action: action,
        headers: headers,
        fin: fin,
    };
}
export function createPushMessage(messageId, action, payload, headers, fin) {
    if (headers === void 0) { headers = Headers(); }
    if (fin === void 0) { fin = true; }
    return {
        kind: MessageKind.Push,
        payload: payload,
        messageId: messageId,
        action: action,
        headers: headers,
        fin: fin,
    };
}
export function createResponseMessage(messageId, status, payload, headers, fin) {
    if (headers === void 0) { headers = Headers(); }
    if (fin === void 0) { fin = true; }
    return {
        kind: MessageKind.Response,
        payload: payload,
        status: status,
        messageId: messageId,
        headers: headers,
        fin: fin,
    };
}
export function createCloseMessage(status, message) {
    return { kind: MessageKind.Close, message: message, status: status };
}
export function createFollowMessage(messageId, payload, fin) {
    if (fin === void 0) { fin = true; }
    return { kind: MessageKind.Follow, messageId: messageId, payload: payload, fin: fin };
}
