// stmp/lib/types.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-11-20 15:35:44
 *
 * shared types for client & server
 * TODO
 */
Object.defineProperty(exports, "__esModule", { value: true });
const is_1 = require("monofile-utilities/lib/is");
var STMPSource;
(function (STMPSource) {
    STMPSource["HANDLE_REQUEST"] = "HANDLE_REQUEST";
    STMPSource["STATUS_NOT_OK"] = "STATUS_NOT_OK";
    STMPSource["REQUEST_TIMEOUT"] = "REQUEST_TIMEOUT";
    STMPSource["TRANSPORT_ERROR"] = "TRANSPORT_ERROR";
    STMPSource["CONNECTION_CLOSED"] = "CONNECTION_CLOSED";
    STMPSource["PARSE_RESPONSE_ERROR"] = "PARSE_RESPONSE_ERROR";
})(STMPSource = exports.STMPSource || (exports.STMPSource = {}));
function newError(source, options = {}) {
    options.source = source;
    if (options.error instanceof Error) {
        Object.setPrototypeOf(options, options.error);
        options.name = `${options.error.name}(${source})`;
    }
    else {
        Object.setPrototypeOf(options, new Error(''));
        options.name = source;
    }
    return options;
}
exports.newError = newError;
function isSTMPError(error) {
    return is_1.isObject(error) && typeof error.source === 'string';
}
exports.isSTMPError = isSTMPError;
